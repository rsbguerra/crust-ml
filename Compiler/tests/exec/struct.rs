struct S {
    i: i32,
    b: bool
}
fn print_bool(b: bool) {
    if b { print!("true") } else { print!("false") }
}

fn print_int(x: i32) {
    if x < 0 { print!("-") }
    else if x > 0 { print!("+") }
    else { print!("0") }
}

fn print_s(s: &S) {
    print!("[ ");
    print_int(s.i);
    print!(", ");
    print_bool(s.b);
    print!(" ]\n");
}

fn foo(s: &mut S) {
    print!("entering foo...\n");
    print_s(s);
    if s.b { s.i = - s.i; }
    s.b = ! s.b;
    print_s(s);
    print!("...leaving foo\n");
}

fn main() {
    let mut s = S { i: 42, b: true };
    foo(&mut s);
    foo(&mut s);
    foo(&mut s);
    foo(&mut s);
}
