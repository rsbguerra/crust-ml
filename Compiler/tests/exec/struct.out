entering foo...
[ +, true ]
[ -, false ]
...leaving foo
entering foo...
[ -, false ]
[ -, true ]
...leaving foo
entering foo...
[ -, true ]
[ +, false ]
...leaving foo
entering foo...
[ +, false ]
[ +, true ]
...leaving foo
